service/kas/access/rewrap.go

```diff
@@ -996,7 +1042,73 @@ func extractNanoPolicy(symmetricKey ocrypto.ProtectedKey, header sdk.NanoTDFHead
 func failAllKaos(reqs []*kaspb.UnsignedRewrapRequest_WithPolicyRequest, results policyKAOResults, err error) {
 	for _, req := range reqs {
 		for _, kao := range req.GetKeyAccessObjects() {
-			failedKAORewrap(results[req.GetPolicy().GetId()], kao, err)
+			failedKAORewrap(results[req.GetPolicy().GetId()].kaoResults, kao, err)
+		}
+	}
+}
+
+// Populate response metadata with required obligations
+func populateRequiredObligationsOnResponse(response *kaspb.RewrapResponse, obligations []string, policyID string) {
+	metadata := response.GetMetadata()
+	if metadata == nil {
+		metadata = make(map[string]*structpb.Value)
+	}
+
+	var fields map[string]*structpb.Value
+	obligationValue, ok := metadata[triggeredObligationsHeader]
+	if !ok || obligationValue.GetStructValue() == nil {
+		fields = make(map[string]*structpb.Value)
+		metadata[triggeredObligationsHeader] = structpb.NewStructValue(&structpb.Struct{
+			Fields: fields,
+		})
+	} else {
+		fields = metadata[triggeredObligationsHeader].GetStructValue().GetFields()
+	}
+
+	values := make([]*structpb.Value, len(obligations))
+	for i, obligation := range obligations {
+		values[i] = structpb.NewStringValue(obligation)
+	}
+	fields[policyID] = structpb.NewListValue(&structpb.ListValue{
+		Values: values,
+	})
+	response.Metadata = metadata
+}
+
+// Retrieve additional request context needed for rewrap processing
+// Header is json encoded AdditionalRewrapContext struct
+// Example: `{"obligations": ["https://demo.com/obl/test/value/watermark","https://demo.com/obl/test/value/geofence"]}`
```

Comment:
This is a good approach, but it can be more extensible if we make `obligations` an object so we can add more keys like `request_flags`, `geospatial`, or general PEP metadata.